[5.x] Word validation error messages more precisely

ravage84 opened the pull request:
Hereby, I propose to improve on the dreaded and very unhelpful `The provided value is invalid` validation error messages.

Wasn't sure whether we can backport this to 4.next, so I implemented it in 5.x to target 5.0.

TODO:
- [x] [There are many more validation methods left](https://github.com/cakephp/cakephp/blob/bd4b99481cfead7d27e876f29291c828b0e3ca04/src/Validation/Validator.php#L1072-L1072)
-  [x] [In case I18n is used, a translated message should be used](https://github.com/cakephp/cakephp/blob/b582d55a3376e9a746f5b904876d0c0fc221ef4e/src/Validation/Validator.php#L2537-L2539)
- [x] Test coverage needs to be updated

ravage84 committed:
Word scalar validation error message more precisely

ravage84 committed:
Word array validation error message more precisely

ravage84 committed:
Translate improved validation error messages if I18n is used

ADmad commented on src/Validation/Validator.php:
```diff
@@ -2184,6 +2184,11 @@ public function integer(string $field, ?string $message = null, Closure|string|n
     {
+        $message ??= 'The provided value must be an array';
+        if ($this->_useI18n) {
+            $message = __d('cake', 'The provided value must be an array');
```
You are potentially overriding user provided value here.

markstory replied:
Yeah we can't overwrite messages like this as we'll replace developer provided messages here.

ravage84 replied:
That wasn't intended. Copy & Pasta.

Which style for checking the user provided message do you prefer?

1. With encompassing  null check:

``` php
        if ($message === null) {
            $message = 'The provided value must be scalar';
            if ($this->_useI18n) {
                $message = __d('cake', 'The provided value must be scalar');
            }
        }
```

2. With null coalescing assignment operators:

``` php
        $message ??= 'The provided value must be scalar';
        if ($this->_useI18n) {
            $message ??= __d('cake', 'The provided value must be scalar');
        }
```

I think the first is more traditional and uses less operations.

3. With encompassing  null check & if/else trending to the use of I18n:

``` php
        if ($message === null) {
            if ($this->_useI18n) {
                $message = __d('cake', 'The provided value must be scalar');
            } else {
                $message = 'The provided value must be scalar';
            }
        }
```

4. With encompassing  null check & if/else trending to the non-use of I18n:

``` php
        if ($message === null) {
            if (!$this->_useI18n) {
                $message = 'The provided value must be scalar';
            } else {
                $message = __d('cake', 'The provided value must be scalar');
            }
        }
```

From an optical view-point (pun wasn't intended) , I would go with the fourth.  For readability the first one, even though it might be a little less efficient (for I18n-enabled applications).

ADmad replied:
```php
         $message ??= $this->_useI18n
                 ? __d('cake', 'The provided value must be scalar')
                 :  'The provided value must be scalar';
```

ravage84 replied:
@ADmad please no multi-line ternary. Especially not in combination with a null coalescing assignment operator (which is something like a ternary in disguise).

Less code doesn't make it more readable.

markstory replied:
If you end up going with a more verbose implementation could we extract it into a protected method so we don't end up with multiple copies. You might need to use validation message templates to make this work. Perhaps a message like `The provided value must be {0}` would work?

ravage84 replied:
> If you end up going with a more verbose implementation could we extract it into a protected method so we don't end up with multiple copies. 

Certainly. Extracting it into a protected method was actually my idea, too.

> You might need to use validation message templates to make this work. Perhaps a message like `The provided value must be {0}` would work?

I'm not sure if using such a split sentence is good from a I18n point of view. In some languages one would probably word the whole sentense differently in some cases instead of just replacing one term with another.
That is why I would go with full sentences.

ravage84 replied:
Went with the last style.

ravage84 committed:
Introduce method to choose between untranslated or translated message

ravage84 commented:
@markstory & @ADmad how does the introduced validation message method look to you?

The only downside to this is that the ``__d()`` method always gets called regardless of whether I18n is used.

ADmad commented on src/Validation/Validator.php:
```diff
@@ -2184,6 +2184,11 @@ public function integer(string $field, ?string $message = null, Closure|string|n
     {
+        $message ??= $this->getValidationMessage(
+            'The provided value must be an array',
+            __d('cake', 'The provided value must be an array')
```
Why not do the translation too inside the helper method instead of repeating the string here?

ravage84 replied:
Because the translation string needs to be found when extracting I18n translation strings.

https://book.cakephp.org/4/en/console-commands/i18n.html#generating-pot-files

As much as I remember we cannot use variables on this method because the strings won't be found that way.

May be I'm not up to date on that, though.

markstory commented on src/Validation/Validator.php:
```diff
@@ -2532,11 +2539,11 @@ protected function _processRules(string $field, ValidationSet $rules, array $dat
-        if ($this->_useI18n) {
-            $message = __d('cake', 'The provided value is invalid');
-        }
+        $message = $this->getValidationMessage(
+            'The provided value is invalid',
+            __d('cake', 'The provided value is invalid')
+        );
```
This will work ok in the fullstack framework where `useI18n` is always true. However, it will create runtime errors for standalone usage of the validation package. We'd need to have the calls to `__d` guarded. The need of guards was why I suggested we use placeholders, but it doesn't sound like that will work as the verb conjugation will depend on the noun gender.

Another option could be to have a case ladder for the i18n messages :shrug: . It wouldn't be simple but validators could use `$message = $this->getValidationMessage('invalid')` and we could have a `match` like:

```php
if ($this->_useI18n) {
  return match ($type) {
    case 'invalid' => __d('cake', 'The provided value is invalid.'),
    case 'array' => __d('cake', 'The provided value is not an array.'),
  }
}

return match ($type)  {
 // non i18n messages here
}
```
It won't be pretty but it will avoid the runtime errors and allow full message translation. :shrug:

ravage84 replied:
I reverted the ``getValidationMessage()`` again as I think using the verbose style is still more readable and guards the calls to `__d`  appropriately.

ravage84 committed:
Remove getValidationMessage() and guard calls to __d()

ravage84 committed:
Correct see annotation

ravage84 committed:
Set individual data validation error messages

ravage84 commented:
Many of the data validation error messages could still be improved, I guess.

Either by using more precise wording. E.g. `The provided value must not be alphanumeric string` than just `The provided value must not be alphanumeric`.

Or some could reflect all configurable options, like `multipleOptions()`.

But the current messages are already far more useful than before.

markstory commented on src/Validation/Validator.php:
```diff
@@ -1420,6 +1552,18 @@ public function equalToField(
+            } else {
+                $message = __d(
+                    'cake',
+                    'The provided value must be equal to the one of field `0`',
```
```suggestion
                    'The provided value must be equal to the one of field `{0}`',
```

markstory commented on src/Validation/Validator.php:
```diff
@@ -1445,6 +1589,18 @@ public function notEqualToField(
+            } else {
+                $message = __d(
+                    'cake',
+                    'The provided value must not be equal to the one of field `0`',
```
```suggestion
                    'The provided value must not be equal to the one of field `{0}`',
```

ravage84 committed:
Interpolate translatioin variables correctly

Co-authored-by: Mark Story <mark@mark-story.com>

ravage84 committed:
Merge branch '5.x' into 5.x-improved-validation-error-messages

ravage84 committed:
Adjust asserted data validation error message

ravage84 committed:
Remove assertion on unset data validation error message

ravage84 committed:
Set notBlank data validation error message to previous text

ravage84 committed:
Assert data validation error message should be set

ravage84 committed:
Fix typo

ravage84 committed:
Fix typos

ravage84 committed:
Fix casing in method call

ravage84 committed:
Cover improved non-i18n validation error messages with assertions

Use newly added NoI18nValidator for these tests.

ravage84 committed:
Also cover remaining getIterator() method

ravage84 committed:
Import Traversable

ravage84 committed:
Fix typo

ravage84 committed:
Assert data validation message without executing validation

We only care about the configured data validation message, not the validation.
It's potentially a little faster, too.

ravage84 committed:
Cover default data validation error message without I18n enabled at least once

ravage84 committed:
Add missing declare strict

ravage84 committed:
CS

ravage84 commented:
The build error in "CI / testsuite (8.3, mysql) (pull_request) " looks unrelated:

> Error: Input required and not supplied: path

https://github.com/cakephp/cakephp/actions/runs/5969787680/job/16196283577?pr=17189#step:11:12

ravage84 commented on src/Validation/Validator.php:
```diff
@@ -1200,6 +1260,14 @@ public function creditCard(
     ) {
+        if ($message === null) {
+            if (!$this->_useI18n) {
+                $message = 'The provided value must be a credit card number';
+            } else {
+                $message = __d('cake', 'The provided value must be a credit card number');
```
```suggestion
                $message = 'The provided value must be a valid credit card number';
            } else {
                $message = __d('cake', 'The provided value must be a valid credit card number');
```

ravage84 commented on src/Validation/Validator.php:
```diff
@@ -1320,6 +1420,14 @@ public function equals(
     ) {
+        if ($message === null) {
+            if (!$this->_useI18n) {
+                $message = sprintf('The provided value must be equals to `%s`', $value);
+            } else {
+                $message = __d('cake', 'The provided value must be equals to `{0}`', $value);
```
```suggestion
                $message = sprintf('The provided value must be equal to `%s`', $value);
            } else {
                $message = __d('cake', 'The provided value must be equal to `{0}`', $value);
```

https://www.quora.com/Which-is-correct-is-equals-to-OR-is-equal-to

ravage84 commented on src/Validation/Validator.php:
```diff
@@ -1344,6 +1452,14 @@ public function notEquals(
     ) {
+        if ($message === null) {
+            if (!$this->_useI18n) {
+                $message = sprintf('The provided value must not be equals to `%s`', $value);
+            } else {
+                $message = __d('cake', 'The provided value must not be equals to `{0}`', $value);
```
```suggestion
                $message = sprintf('The provided value must not be equal to `%s`', $value);
            } else {
                $message = __d('cake', 'The provided value must not be equal to `{0}`', $value);
```

https://www.quora.com/Which-is-correct-is-equals-to-OR-is-equal-to

ravage84 commented on tests/TestCase/Validation/ValidatorTest.php:
```diff
@@ -1918,6 +1980,12 @@ public function testCreditCard(): void
+
+        $fieldName = 'field_name';
+        $rule = 'creditCard';
+        $expectedMessage = 'The provided value must be a credit card number';
```
```suggestion
        $expectedMessage = 'The provided value must be a valid credit card number';
```

ravage84 committed:
Improve wording of credit card data validatoin error message

ravage84 committed:
Equal to, not equals to

https://www.quora.com/Which-is-correct-is-equals-to-OR-is-equal-to

ravage84 committed:
Adjust assertion for equal to validation error messages

ravage84 committed:
Mention the number of decimal places in decimal data validation error message

ravage84 committed:
Pass additional data even when it's an empty array

ravage84 committed:
Mention the valid options in the inList data validation error message

ravage84 committed:
Allow credit card types be given as array

This was already supported by the validation method itself.

ravage84 committed:
Mention the valid card types in the credit card data validation error message

ravage84 committed:
Pass additional data to non-i18n validator only when not null

ravage84 committed:
Mention the valid formats in the date/time data validation error message

ravage84 committed:
Mention the valid formats in the date data validation error message

ravage84 commented on src/Validation/Validator.php:
```diff
@@ -2247,6 +2731,13 @@ public function multipleOptions(
         ?string $message = null,
         Closure|string|null $when = null
     ) {
+        if ($message === null) {
+            $message = 'The provided value must be a set of multiple options';
+            if ($this->_useI18n) {
+                $message = __d('cake', 'The provided value must be a set of multiple options');
+            }
+        }
```
This could mention the options set.:

https://github.com/cakephp/cakephp/blob/332baf6a13532e60c5537315edfdfe41c1a68eee/src/Validation/Validation.php#L944-L946

But this would lead to many code branches. Not sure if that's worth the effort.

ravage84 committed:
Cover empty enumeration for date, date/time & credit card data validation

ravage84 commented:
> Many of the data validation error messages could still be improved, I guess.
> [...]
> Or some could reflect all configurable options, like `multipleOptions()`.

Done in most cases.

ravage84 commented:
@markstory can we backport this to 4.5.0?

It could lead to breaking test cases in user land applications if they assert the data validation error messages in their test suite but apart from that the improved wording should be seen as improvement from a UX perspective.

If so, I'm going to do that.
